ComponentKit/Core/Scope/CKComponentScopeHandle.mm

```diff
@@ -125,12 +142,28 @@ - (BOOL)acquireFromComponent:(CKComponent *)component
 {
   if (!_acquired && [component isMemberOfClass:_componentClass]) {
     _acquired = YES;
+    _acquiredComponent = component;
     return YES;
   } else {
     return NO;
   }
 }
 
+- (void)resolve
+{
+  if (!_controller) {
+    CKThreadLocalComponentScope *currentScope = CKThreadLocalComponentScope::currentScope();
+    CKAssert(currentScope != nullptr, @"Current scope should never be null here. Thread-local stack is corrupted.");
+
+    // A controller can be non-nil at this callsite during component re-generation because a new scope handle is
+    // generated in a new tree, that is acquired by a new component. We pass in the original component controller
+    // in that case, and we should avoid re-generating a new controller in that case.
+    _controller = newController(_acquiredComponent, currentScope->newScopeRoot);
+  }
+  _acquiredComponent = nil;
```

Comment:
I'd just leave a comment here explaining this is where `CKComponentScopeHandle` explicitly breaks the purposeful retain cycle.